File: programmers/etc/level3/부대복귀.py
from collections import deque

def solution(n, roads, sources, destination):
    graph = [[] for _ in range(n+1)]
    for a, b in roads:
        graph[a].append(b)
        graph[b].append(a)
        
    visited = [-1]*(n+1)
    visited[destination] = 0

    q = deque([destination])
    
    while q:
        cur = q.popleft()
        for nxt in graph[cur]:
            if visited[nxt] == -1:
                visited[nxt] = visited[cur]+1
                q.append(nxt)
    
    return [visited[s] for s in sources]


n = 3
roads = [[1, 2], [2, 3]]
sources = [2, 3]
destination = 1
print(solution(n, roads, sources, destination)) # [1, 2]

n = 5
roads = [[1, 2], [1, 4], [2, 4], [2, 5], [4, 5]]
sources = [1, 3, 5]
destination = 5
print(solution(n, roads, sources, destination)) # [2, -1, 0]

File: Java/programmers/etc/level3/부대복귀.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 부대복귀 {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        
        List<List<Integer>> graph = new ArrayList<>();
        for (int i=0; i<=n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] road : roads) {
            graph.get(road[0]).add(road[1]);
            graph.get(road[1]).add(road[0]);
        }

        Queue<Integer> q = new LinkedList<>();
        int[] visited = new int[n+1];
        Arrays.fill(visited, -1);
        
        q.add(destination);
        visited[destination] = 0;

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (Integer nxt : graph.get(cur)) {
                if (visited[nxt] == -1) {
                    visited[nxt] = visited[cur]+1;
                    q.add(nxt);
                }
            }
        }

        int idx = 0;
        for (int source : sources) {
            answer[idx++] = visited[source];
        }

        return answer;
    }
}
